allow extensions to StandardMaterial

robtfm opened the pull request:
# Objective

allow extending `Material`s (including the built in `StandardMaterial`) with custom vertex/fragment shaders and additional data, to easily get pbr lighting with custom modifications, or otherwise extend a base material. 

# Solution

- added `ExtendedMaterial<B: Material, E: MaterialExtension>` which contains a base material and a user-defined extension.
- added example `extended_material` showing how to use it
- modified AsBindGroup to have "unprepared" functions that return raw resources / layout entries so that the extended material can combine them

note: doesn't currently work with array resources, as i can't figure out how to make the OwnedBindingResource::get_binding() work, as wgpu requires a `&'a[&'a TextureView]` and i have a `Vec<TextureView>`. 

# Migration Guide

manual implementations of `AsBindGroup` will need to be adjusted, the changes are pretty straightforward and can be seen in the diff for e.g. the `texture_binding_array` example.

james7132 commented:
Not sure if the PR description is not using the same formatting the CI check expects, but please ignore that comment.

JMS55 commented:
Does this provide an easy way to do something like "use shader code to generate a random color, feed that into the base color of the pbr shader"? Or can we only modify the output, not the input?

robtfm commented:
> Does this provide an easy way to do something like "use shader code to generate a random color, feed that into the base color of the pbr shader"? Or can we only modify the output, not the input?

you can modify the vertex output / pbr_fragment input, so for that particular case i guess you could use vertex colors and override the generated vertex color. 

you can't more generally change the material bindings since the pbr_fragment function reads them directly. for that you would have to go one level deeper and construct a `PbrInput` and use the pbr function directly (you'd also have to duplicate the tonemapping, fog etc as well)... 

I could introduce another function so that there's an entry point which takes a PbrInput and calls pbr plus does all the in-shader postprocessing as well. i think that sounds like a good idea.

JMS55 commented:
> you can't more generally change the material bindings since the pbr_fragment function reads them directly. for that you would have to go one level deeper and construct a PbrInput and use the pbr function directly (you'd also have to duplicate the tonemapping, fog etc as well)...

I could introduce another function so that there's an entry point which takes a PbrInput and calls pbr plus does all the in-shader postprocessing as well. i think that sounds like a good idea.

Yeah, exactly. I'm thinking of the use case where you want to make procedurally generated materials, but have them work with PBR lighting with proper material information. Like how blender lets you build a shader graph with inputs for the PBR parameters.

NotAFile commented:
> you can modify the vertex output / pbr_fragment input

What would be some use cases for modifying the outputs there as opposed to the inputs?

I can see use cases for hooking in before the vertex shader (e.g. heightmapping) and before the fragment shader (e.g. animated texture) but I can't immediately think of anything I would want do that would come at those stages specifically.

robtfm commented:
> What would be some use cases for modifying the outputs there as opposed to the inputs?

vertex output is fragment input, so this is the right place to do animated textures if you want smooth non-linear adjustments between vertexes.

if you meant the fragment output, in the example i modified the output to do some kind of rough cell-shading type thing. i imagine some other mesh-specific (not full screen) post-processing could also fit here.

prime31 commented:
This looks fantastic! 2 👍's up

robtfm commented:
> Theoretically it would also make sense to extend a `Material2d`

that's theoretically true, but in practice it's easy enough to make your own material for 2d.

The built in 2d frag shader is only 10 lines long so you can copy/paste it without much drama. `pbr.wgsl`'s frag shader is more than a thousand lines over a dozen files so it's hard to copy/paste and keep in sync over bevy versions, etc, so this extension is more useful there.

coreh commented:
> I could introduce another function so that there's an entry point which takes a PbrInput and calls pbr plus does all the in-shader postprocessing as well. i think that sounds like a good idea.

Yeah, this could (eventually) enable all sorts of interesting effects, such as decals and procedural water surfaces with parallax mapping.

Maybe we could also have a “inversion of control” pattern, where instead of the extended shader owning the “true” entrypoint, there are multiple “hook”/"override" points that are called back by the main PBR shader in specific places, for example:

- After preparing `PbrInput` (your proposal above)
- After fetching each light
- After fetching each light's shadows (could be useful for creating “fake” shadows, like in some 3d platformers)
- Before combining direct / indirect lights, etc
- After the PBR output is produced (the one in the current PR)

fintelia commented:
Isn't this going to behave differently depending on whether tone mapping happens in the PBR shader (non-HDR camera) or in a separate pass later (HDR camera)?

robtfm commented:
> Isn't this going to behave differently depending on whether tone mapping happens in the PBR shader (non-HDR camera) or in a separate pass later (HDR camera)?

yes it will. i'm not sure that's a huge issue since i would expect the user to use hdr or not exclusively (not mix them and expect the same results). and if necessary the shader defs could be used in the top-level frag shader as well to alter the behaviour.

but i do agree that it would be nice to inject user functionality before the "post-processing" section of the main shader sequence (fog, debanding, etc as well as tonemapping). that, plus jasmine's suggestion about altering inputs, suggests that maybe we should have a handful of core pbr functions (`create_pbr_input`, `apply_lighting`, `apply_processing`) that the standard shader calls in order, and the user could pick and mix / modify intermediate results as appropriate.

i'll try and bring this up to date and try that approach as well in the next week or so.

cart commented:
Moving to 0.12. Seems useful but needs more reviews (and I'd like to consider the design a bit).

Malax commented on assets/shaders/quantize_shader.wgsl:
```diff
@@ -0,0 +1,18 @@
+var<uniform> my_extended_material: MyExtendedMaterial;
+
+@fragment
+fn fragment(in: FragmentInput) -> @location(0) vec4<f32> {
+    // call to the standard pbr fragment shader
+    var output_color = pbr_fragment(in);
+
+    // we can then modify the results using the extended material data
+    output_color = vec4<f32>(vec4<u32>(output_color * f32(my_extended_material.quantize_steps))) / f32(my_extended_material.quantize_steps);
+    return output_color;
+}
```
Looking at this example, does this mean there is no way to, for example, set the color _before_ the shadows are added on top?

robtfm replied:
kind of, see the [comment above](https://github.com/bevyengine/bevy/pull/7820#issuecomment-1472349197) plus the next couple. i will amend it so you can generate a PbrInput struct with a first function call, modify it, then pass it on to lighting / postprocessing with separate functions.

robtfm commented:
i updated this a couple of weeks ago but forgot to comment. i added the ability to intercept and modify the pbrInput struct before lighting is applied, and ability to modify before (or skip) in-shader post-processing (fog, tonemapping, debanding).

it doesn't allow for more fine-grained intercepting (per light, e.g). i'm not sure there's a good way to do that without using virtual functions, which would definitely be a separate pr.

coreh commented on crates/bevy_pbr/src/extended_material.rs:
```diff
@@ -0,0 +1,178 @@
+    }
+
+    fn alpha_mode(&self) -> crate::AlphaMode {
+        StandardMaterial::alpha_mode(&self.standard)
```
I can think of a few scenarios where it would make sense to have the extended material control the `AlphaMode`, especially for effects that conditionally use `AlphaMode::Mask`, like screen door transparency for fading between different impostor levels, or for “despawning” or “dissolving” effects.

This is a relatively straightforward rule I think could work for enabling standard material to still control the alpha mode for simple extended materials, but still give extended material the ultimate control.

```suggestion
        let standard = StandardMaterial::alpha_mode(&self.standard);
        let extended = M::alpha_mode(&self.extended);

        if extended == crate::AlphaMode::Opaque {
            // When extended alpha mode is opaque, use standard's alpha mode
            standard
        } else {
            // Otherwise, extended will override standard
            extended
        }
```

<img src="https://github.com/bevyengine/bevy/assets/418473/6462e166-b7dd-436e-9ba5-e52e48a04bdf"  alt="screen door transparency">
<img src="https://github.com/bevyengine/bevy/assets/418473/9d3a9483-fd4d-42ae-8dfe-92d5d16eee63"  alt="dissolve effect">

robtfm replied:
...whereas here it's not quite so obvious how to combine them. using opaque as default makes sense, but isn't entirely obvious.

i wonder if it would make sense to add another `MaterialExtension` trait that doesn't include depth bias or blend mode, and require that (`ExtendedMaterial<M: MaterialExtension>`)

robtfm replied:
>`MaterialExtension` trait

i added this, i think it's better but let me know what you think.

robtfm replied:
i just realised now that this doesn't give you control over the mode using the ExtendedMaterial - so it's no longer ambiguous but you'd still have to access the `StandardMaterial` asset in order to modify the blend mode. I definitely prefer only having one way to do it though, even if it means needing an extra asset lookup to modify the setting ...

superdump replied:
Independent of this PR, I've felt like AlphaMode is about RenderPhases and isn't specific to StandardMaterial. It should be possible to set in 2D as well.

superdump replied:
And yeah, having one way to configure alpha mode is much clearer imo.

coreh replied:
Hmm yeah if we eventually extract `AlphaMode` as its own component, then you can handle this with a system that just reads from whatever extended material you want and writes to `AlphaMode`, and it's both conceptually cleaner and fully consistent.

coreh commented on crates/bevy_pbr/src/lib.rs:
```diff
@@ -136,6 +139,12 @@ impl Plugin for PbrPlugin {
+        load_internal_asset!(
+            app,
+            PBR_FRAGMENT_HANDLE,
+            "render/pbr_fragment.wgsl",
```
Does the name `pbr_fragment` mean something like “use this in the fragment shader for pbr-based materials”? Do we need a separate file for this? Would the contents not make sense under the existing `pbr.wgsl` file?

robtfm replied:
i chose to separate them mainly to make the reusable `pbr_fragment` api clearer. i do agree the actual `pbr.wgsl` is pretty basic now, but it feels nicer to me this way, and arguably makes for an easier "first step" into the shader code for people browsing the source. 

including it also potentially blocks allowing imported modules to specify shader entry points in future (though naga_oil doesn't support imported entry points anyway right now, i think it could do in future without too much effort).

i'd rather keep this as is unless there's a concensus against it.

coreh reviewed (commented):
Really like this, overall it's much more powerful when compared to the previous version that I tried out, and I think it strikes a good balance between allowing customization at three very important separate stages (before pbr, between pbr and post processing, and after post processing) while not making it super hard to use in order to support further customizations.

Left some comments re: naming of some of the WGSL functions/files, and also a suggestion re: the `Material` impl for `ExtendedMaterial<M>` for alpha mode. Can't comment on the procedural macros changes, as I don't have a lot of experience with those yet.

Tested and it works well on my machine (Apple Silicon)

robtfm commented on crates/bevy_pbr/src/extended_material.rs:
```diff
@@ -0,0 +1,178 @@
+    }
+
+    fn depth_bias(&self) -> f32 {
+        M::depth_bias(&self.extended) + StandardMaterial::depth_bias(&self.standard)
```
yes, i chose to do it here because it was obvious and easy to combine them...

superdump commented on crates/bevy_pbr/src/render/pbr_fragment.wgsl:
```diff
@@ -0,0 +1,185 @@
+    let double_sided = (pbr_bindings::material.flags & pbr_types::STANDARD_MATERIAL_FLAGS_DOUBLE_SIDED_BIT) != 0u;
+
+    var pbr_input: pbr_functions::PbrInput = pbr_input_from_mesh_vertex_output(in, is_front, double_sided);
+    pbr_input.material.base_color = pbr_input.material.base_color * pbr_bindings::material.base_color;
```
```suggestion
    pbr_input.material.base_color *= pbr_bindings::material.base_color;
```

superdump commented on crates/bevy_pbr/src/render/pbr_fragment.wgsl:
```diff
@@ -0,0 +1,185 @@
+
+#ifdef VERTEX_UVS
+    if ((pbr_bindings::material.flags & pbr_types::STANDARD_MATERIAL_FLAGS_BASE_COLOR_TEXTURE_BIT) != 0u) {
+        pbr_input.material.base_color = pbr_input.material.base_color * textureSampleBias(pbr_bindings::base_color_texture, pbr_bindings::base_color_sampler, uv, view.mip_bias);
```
```suggestion
        pbr_input.material.base_color *= textureSampleBias(pbr_bindings::base_color_texture, pbr_bindings::base_color_sampler, uv, view.mip_bias);
```

superdump commented on crates/bevy_pbr/src/render/pbr_fragment.wgsl:
```diff
@@ -0,0 +1,185 @@
+        if ((pbr_bindings::material.flags & pbr_types::STANDARD_MATERIAL_FLAGS_METALLIC_ROUGHNESS_TEXTURE_BIT) != 0u) {
+            let metallic_roughness = textureSampleBias(pbr_bindings::metallic_roughness_texture, pbr_bindings::metallic_roughness_sampler, uv, view.mip_bias);
+            // Sampling from GLTF standard channels for now
+            metallic = metallic * metallic_roughness.b;
```
```suggestion
            metallic *= metallic_roughness.b;
```

superdump commented on crates/bevy_pbr/src/render/pbr_fragment.wgsl:
```diff
@@ -0,0 +1,185 @@
+            let metallic_roughness = textureSampleBias(pbr_bindings::metallic_roughness_texture, pbr_bindings::metallic_roughness_sampler, uv, view.mip_bias);
+            // Sampling from GLTF standard channels for now
+            metallic = metallic * metallic_roughness.b;
+            perceptual_roughness = perceptual_roughness * metallic_roughness.g;
```
```suggestion
            perceptual_roughness *= metallic_roughness.g;
```

superdump commented on crates/bevy_pbr/src/render/pbr_fragment.wgsl:
```diff
@@ -0,0 +1,185 @@
+    output_color = pbr_functions::premultiply_alpha(pbr_input.flags, output_color);
+#endif
+    return output_color;
+}
```
Missing newline at the end of the file.

superdump commented on crates/bevy_pbr/src/render/pbr_fragment.wgsl:
```diff
@@ -0,0 +1,185 @@
+#ifdef DEBAND_DITHER
+    var output_rgb = output_color.rgb;
+    output_rgb = powsafe(output_rgb, 1.0 / 2.2);
+    output_rgb = output_rgb + screen_space_dither(pbr_input.frag_coord.xy);
```
```suggestion
    output_rgb += screen_space_dither(pbr_input.frag_coord.xy);
```

superdump commented on crates/bevy_render/src/render_resource/bind_group.rs:
```diff
@@ -267,17 +270,68 @@ pub trait AsBindGroup {
+        })
+    }
+
+    /// Returns a vec of (binding index, `OwnedBindingResource`).
```
Is there any reason a `Vec<OwnedBindingResource>` with binding indices as the positions in the vec is not sufficient?

robtfm replied:
yes - binding indices don't have to start from zero

superdump reviewed (commented):
Mostly nits.

I really like this. Great stuff!

AsBindGroup should imo not create buffers (my material data in arrays in buffers stuff will take care of that) and not create bind groups, as in this PR. So perhaps it needs renaming to `AsBindable` or something?

JMS55 commented:
> AsBindGroup should imo not create buffers (my material data in arrays in buffers stuff will take care of that) and not create bind groups, as in this PR. So perhaps it needs renaming to AsBindable or something?

Agreed, but lets leave it to a future PR (after 0.12 releases).

JMS55 commented on assets/shaders/extended_material.wgsl:
```diff
@@ -0,0 +1,36 @@
+    // we can optionally modify the lit color before post-processing is applied
+    let modified_lit_color = vec4<f32>(vec4<u32>(lit_color * f32(my_extended_material.quantize_steps))) / f32(my_extended_material.quantize_steps);
+
+    // apply in-shader post processing (fog, tonemapping, debanding)
```
Needs a callout here on tonemapping/debanding (idt fog is included in that) being done in the TonemappingNode when Camera::hdr is true, instead of in the material.

JMS55 commented on crates/bevy_pbr/src/extended_material.rs:
```diff
@@ -0,0 +1,201 @@
+
+use crate::{Material, MaterialPipeline, MaterialPipelineKey, StandardMaterial};
+
+/// A subset of the `Material` trait for use when extending `StandardMaterial`s.
```
"for use when" awkward warning, idk what would be better though.

robtfm replied:
"for defining extensions to" ?

JMS55 replied:
yeah that works

JMS55 commented on crates/bevy_pbr/src/extended_material.rs:
```diff
@@ -0,0 +1,201 @@
+/// If the material `M` returns a non-default result from `vertex_shader()` it will be used in place of the standard
+/// vertex shader (`bevy_pbr::render::mesh.wgsl`). It should return the same output as the standard vertex shader.
+/// If the material `M` returns a non-default result from `fragment_shader()` it will be used in place of the standard
+/// fragment shader (`bevy_pbr::render::pbr_fragment.wgsl`). since all the standard material fields are
```
```suggestion
/// fragment shader (`bevy_pbr::render::pbr_fragment.wgsl`). Since all the standard material fields are
```

JMS55 commented on crates/bevy_pbr/src/extended_material.rs:
```diff
@@ -0,0 +1,201 @@
+/// present, the `pbr_fragment` shader function can be called from the custom shader (see
+/// the `extended_material` example).
+#[derive(Asset, Clone, TypePath)]
+pub struct ExtendedMaterial<M: StandardMaterialExtension> {
```
I feel like we should call it `ExtendedStandardMaterial`, as it only works with StandardMaterial (it would be nice if it was generic...).

robtfm replied:
bit of a mouthful but you're right.

i thought generic wouldn't be possible due to orphan rules but now i can't see why. i'm going to have another quick go at that tonight.

robtfm replied:
made it generic

superdump replied:
I agree and had the same thought, but felt `ExtendedMaterial` is maybe implicitly 'more' than `StandardMaterial`. You make a good point that as it only works on top of `StandardMaterial`, it would be better-named `ExtendedStandardMaterial`.

robtfm replied:
it's no longer true - the extension can extend any material now, so i'm leaving this as is

JMS55 commented on crates/bevy_pbr/src/extended_material.rs:
```diff
@@ -0,0 +1,201 @@
+
+use crate::{Material, MaterialPipeline, MaterialPipelineKey, StandardMaterial};
+
+/// A subset of the `Material` trait for use when extending `StandardMaterial`s.
```
Please link to `ExtendedMaterial` in the doc comment.

JMS55 commented on crates/bevy_pbr/src/render/pbr_fragment.wgsl:
```diff
@@ -0,0 +1,185 @@
+    return pbr_input;
+}
+
+fn in_shader_post_processing(
```
I don't love using the term post processing, as it has overlap with things like bloom, taa, etc. Maybe call it `in_shader_post_lighting_adjustment` or something?

robtfm replied:
that's why it's "in-shader" post processing. given that 2 of these things are normally done in actual post-processing i think the name's fair.

JMS55 replied:
I'm worried users will think that means bloom is done in the shader here, for instance.

robtfm replied:
added clarifications in the example and the base pbr shader, as well as on the function

superdump replied:
I didn't really like the `in_shader` part when reading it, but couldn't come up with anything I liked better at the time. I was thinking something like `main_pass_...` so maybe a combination like `main_pass_post_lighting_processing`?

robtfm replied:
done

JMS55 commented on assets/shaders/extended_material.wgsl:
```diff
@@ -0,0 +1,36 @@
+    quantize_steps: u32,
+}
+
+@group(1) @binding(100)
```
I don't love telling people to use 100. I feel like we should
* Pick a smaller number (50?). Ideally we could make it a constant and shader def and have people do `$X + 1, $X + 2, etc`
* Give a specific callout on why we're using 100 here (to not overlap with the standard material bindings, which may change between bevy versions). Also add it in the example rust code.

robtfm replied:
> `$X + 1, $X + 2, etc`
this doesn't work unfortunately. we can make a shader def for the offset but wgsl doesn't allow `@binding(100 + 1)`. this will hopefully be resolved at some point by auto bindings, but definitely not for 0.12 (and probably not for 0.13 - i hope to get the view group done for 0.13 as a start).
>Give a specific callout on why we're using 100 here
👍

JMS55 replied:
Ah yeah I forgot to mention that we _can't_ do it yet because naga doesn't allow constant math like that yet.

robtfm replied:
if we don't move to auto bindings then yeah, but not as part of this pr ...

robtfm replied:
also just to note an additional complexity - extensions can be chained / applied on top of one another so the user base isn't foolproof either.

JMS55 reviewed (changes requested):
Overall I'm a fan of the approach :). I've also seen requests for modifying specific parts of the lighting, like custom BSDFs or modifying specific lights. Lets leave that to a future PR though, after this one has had time to settle.

---

Migration guide could be cleaned up a bit, and should link to the diff once this PR is merged.

superdump commented on crates/bevy_render/src/render_resource/bind_group.rs:
```diff
@@ -262,17 +265,68 @@ pub trait AsBindGroup {
+    }
+
+    /// Returns a vec of (binding index, `OwnedBindingResource`).
+    /// In cases where `OwnedBindingResource` is not available (as for array textures currently),
```
```suggestion
    /// In cases where `OwnedBindingResource` is not available (as for bindless texture arrays currently),
```

superdump commented on examples/shader/texture_binding_array.rs:
```diff
@@ -141,36 +141,45 @@ impl AsBindGroup for BindlessMaterial {
+        _: &FallbackImage,
+    ) -> Result<UnpreparedBindGroup<Self::Data>, AsBindGroupError> {
+        // we implement as_bind_group directly because
+        panic!("texture arrays can't be owned")
```
```suggestion
        panic!("bindless texture arrays can't be owned")
```

superdump commented on crates/bevy_pbr/src/render/pbr.wgsl:
```diff
@@ -1,226 +1,43 @@
-            is_front,
-        );
+    // alpha discard
+    pbr_input.material.base_color = alpha_discard(pbr_input.material, pbr_input.material.base_color);
```
robtfm pointed out that this is not a regression. I am glad that when code moves around, I see it with fresh eyes. And, as this is not a regression, this is not a blocking comment.

robtfm replied:
as we discussed, it's a little involved to calculate the gradient and include the mip_bias correctly, so i'd prefer to leave it for future

DGriffin91 replied:
This may not be a regression here, but before https://github.com/bevyengine/bevy/pull/10105 the discard happened much earlier https://github.com/bevyengine/bevy/blob/a15d1526358a3e5fd0ac76339f0c23612842167b/crates/bevy_pbr/src/render/pbr_functions.wgsl#L155 But in https://github.com/bevyengine/bevy/pull/10105 it was also mentioned "it must be done after all texture samples for uniform control flow"

robtfm replied:
the discard was previously in `pbr_functions::pbr()` which was well after all the texture samples (which are in `pbr::fragment()`)

cart commented on examples/shader/extended_material.rs:
```diff
@@ -0,0 +1,83 @@
+    App::new()
+        .add_plugins(DefaultPlugins)
+        .add_plugins(MaterialPlugin::<
+            ExtendedMaterial<StandardMaterial, MyExtension>,
```
Given that extending materials is likely going to be very common in practice, I think its worth considering making it a "first class" aspect of the core Material trait. Using the `extended_material.rs` as an example:

```rust
fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_plugins(MaterialPlugin::<QuantizedStandardMaterial>::default())
        .run();
}

fn setup(
    mut commands: Commands,
    mut materials: ResMut<Assets<QuantizedStandardMaterial>>,
) {
    // sphere
    commands.spawn(MaterialMeshBundle {
        material: materials.add(QuantizedStandardMaterial {
            base: StandardMaterial {
                base_color: Color::RED,
                opaque_render_method: OpaqueRendererMethod::Auto,
                ..Default::default()
            },
            quantize_steps: 10,
        }),
        ..default()
    });
}

#[derive(Asset, AsBindGroup, TypePath, Debug, Clone)]
struct QuantizedStandardMaterial {
    base: StandardMaterial,
    #[uniform(100)]
    quantize_steps: u32,
}

impl Material for QuantizedStandardMaterial {
    type Base = StandardMaterial;
    fn fragment_shader() -> ShaderRef {
        "shaders/extended_material.wgsl".into()
    }
}

impl GetBaseMaterial<StandardMaterial> for QuantizedStandardMaterial {
    fn get_base_material(&self) -> &StandardMaterial {
        &self.base
    }
}
```

This improves a number of things ergonomically. It removes the need for the ExtendedMaterial wrapper when referencing the material type:

```rust
// This
Assets<QuantizedStandardMaterial>
// Versus this
Assets<ExtendedMaterial<StandardMaterial, QuantizedStandardMaterial>>
```
And also removes both the wrapper type and nesting when users define the material:

```rust
// This
QuantizedStandardMaterial {
    base: StandardMaterial {
        base_color: Color::RED,
        opaque_render_method: OpaqueRendererMethod::Auto,
        ..Default::default()
    },
    quantize_steps: 3,
}

// Versus this
ExtendedMaterial {
    base: StandardMaterial {
        base_color: Color::RED,
        opaque_render_method: OpaqueRendererMethod::Auto,
        ..Default::default()
    },
    extension: QuantizedStandardMaterial {
        quantize_steps: 3
    },
}
```

I've verified that this is expressible in the type system. This does have the downside of requiring this for "non-extended" materials:

```rust
impl Material for MyMaterial {
    type Base = ();
    fn fragment_shader() -> ShaderRef {
        "foo.wgsl".into()
    }
}
``` 

This UX hiccup would be resolved by "associated type defaults" (https://github.com/rust-lang/rust/issues/29661). Although progress on this feature appears to have stalled.

We _could_ also move the associated type to a normal generic:

```rust
trait Material<Base = ()> { }
```

Which enables this:

```rust
impl Material for MyMaterial {
    fn fragment_shader() -> ShaderRef {
        "foo.wgsl".into()
    }
}

impl Material<StandardMaterial> for QuantizedStandardMaterial {
    fn fragment_shader() -> ShaderRef {
        "shaders/extended_material.wgsl".into()
    }
}
``` 

But for _Rust reasons_, using a normal generic would mean that you need to re-specify the base material whenever you create a type that needs to support base materials (by explicitly naming `Material<T>`):

```rust
.add_plugins(MaterialPlugin::<QuantizedStandardMaterial, StandardMaterial>::default())
```

cart replied:
Note that `GetBaseMaterial` is a separate trait (rather than existing on `Material`) in the interest of making `impl Material for MyMaterial` not need to supply a `get_base_material(&self) -> &()` method. With a separate trait, we can use a blanket impl on `T` to auto-impl this.

superdump replied:
I like the suggestions. I think whether we merge or not then depends on what time @robtfm has available to implement and test it.

robtfm replied:
one issue that i see with this approach - the extension is not a `Material` but a `MaterialExtension`, and `ExtendedMaterial` works by applying `Material + MaterialExtension => Material`. The base material defines the alpha mode, opaque render method and depth bias, while the extension only defines the shaders (plus data). 

So we would need to somehow say that `E: Material` (rather than `E: MaterialExtension` is required if `Base = ()`, or change it to just use a single material trait and introduce ambiguity on the depth_bias (easily solved by adding), alpha_mode and default opaque method (not easily solved but could be done by convention).

i'll think a bit more about this later, there might be a simple way to make it fit together that i'm missing.

cart replied:
One solution would be to make every "material value" return an `Option` (and default to returning None). Slightly more boilerplate when defining them, but these properties are reasonably uncommon to set manually.

cart replied:
Another major downside to this strategy is that when reading a material value, you need to first check the current material, then its base, etc, etc. Given that this supports infinite nesting, we'd definitely want another set of methods to do this for us.

But supporting multiple nesting cleanly is another major benefit to this approach. The ergonomics of using `ExtendedMaterial` more than once is prohibitive:

```rust
// ExtendedMaterial
Assets<ExtendedMaterial<ExtendedMaterial<ExtendedMaterial, A>, B>, C>>
// Unified Material
Assets<C>
```

robtfm replied:
I had the same thought about returning options. It’s not terrible.

Regarding chaining - using associated types will make the chain parent type fixed for a given extension. I guess it’s not a huge problem since the shader needs to have precise awareness of the parent anyway - at least right now, I was hoping to improve this in future though. 

The long type can be somewhat improved by using a typedef. 

But I guess overall it does seem better to use associated type. I’ll have a go

cart replied:
Arg so sadly now that I've gotten into the details of this impl, I've realized that I missed an important piece:

This is a statically recursive type. For the Material trait (with `Base` associated type) alone, this is expressible. And I believe we could even make it work by checking if the TypeId is `()` and breaking the "material property resolve" recursion when we hit it (ugly but I do think it would _work_).

However this _will not work_ (and is not expressible) for the specialized pipeline key (because actual types cannot recurse like this).

Sadly that means that (with the current specialization system), I don't think we can do a unified trait.

superdump replied:
Too many layers would probably also hurt specialisation performance as I guess the keys would grow and grow.

cart replied:
I'm going to explore one more plan that defines an ExtendedMaterial trait, but implements Material for it.

cart replied:
Ok that plan hits a similar issue where we need the extending material type to implement AsBindGroup, but we need _that_ AsBindGroup impl to return the combined bind group. I think that might be possible by adding "nesting" to the AsBindGroup derive. But we're entering "scope too big to make the change now" territory. I think we should roll with the impl in this PR for now.

cart reviewed (approved):
Massive usability improvement and I do think this is on the right path. I do have one idea to make extending feel a bit more natural and improve the ergonomics of using extended materials. I'm approving the PR (in its current state) though because I see no reason to block progress on this discussion, but if we have time to discuss, I think its worth doing that _before_ merging, as the changes wouldn't be significant.
